src/jdk.javadoc/share/classes/jdk/javadoc/internal/doclets/formats/html/HtmlDocletWriter.java

```diff
@@ -1258,10 +1376,13 @@ public ContentBuilder add(CharSequence text) {
         configuration.tagletManager.checkTags(element, trees);
         commentRemoved = false;
 
-        for (ListIterator<? extends DocTree> iterator = trees.listIterator(); iterator.hasNext();) {
+        var useMarkdown = trees.stream().anyMatch(t -> t.getKind() == Kind.MARKDOWN);
+        var markdownHandler = useMarkdown ? new MarkdownHandler(element) : null;
```

Comment:
The `MarkdownHandler` and `HeadingNodeRenderer` classes must become aware of the current `TagletWriter.Context`.  That's because headings and other block tags should only be generated if `TagletWriter.Context.isFirstSentence` is `false`.